File: nodejs/src/approval.ts
import readline from "node:readline";
import type { ApprovalFn } from "./types.js";

export const promptCliApproval: ApprovalFn = async ({
  executionId,
  agentId,
  action,
  resource,
  summary,
  timeoutSeconds,
}) => {
  const lines = [
    "[sqube] Approval required",
    `  execution_id: ${executionId}`,
    `  agent_id:     ${agentId}`,
    `  action:       ${action}`,
    `  resource:     ${resource ?? ""}`,
    `  summary:      ${summary}`,
    "",
    "Approve? [y/N]",
  ];
  process.stderr.write(lines.join("\n") + "\n");

  return new Promise((resolve) => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stderr });
    let settled = false;

    const timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        rl.close();
        resolve({ approved: false, approvedBy: null, reason: "timeout" });
      }
    }, timeoutSeconds * 1000);

    rl.question("", (answer) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      rl.close();
      const normalized = answer.trim().toLowerCase();
      if (normalized === "y" || normalized === "yes") {
        resolve({ approved: true, approvedBy: "cli_user", reason: null });
      } else {
        resolve({ approved: false, approvedBy: null, reason: "denied_by_user" });
      }
    });
  });
};

File: rust/src/approval.rs
use std::io::{self, Write};

pub struct ApprovalRequest {
    pub execution_id: String,
    pub agent_id: String,
    pub action: String,
    pub resource: Option<String>,
    pub summary: String,
}

pub fn prompt_cli_approval(req: &ApprovalRequest) -> (bool, Option<String>, Option<String>) {
    eprintln!("[sqube] Approval required");
    eprintln!("  execution_id: {}", req.execution_id);
    eprintln!("  agent_id:     {}", req.agent_id);
    eprintln!("  action:       {}", req.action);
    eprintln!("  resource:     {}", req.resource.as_deref().unwrap_or(""));
    eprintln!("  summary:      {}", req.summary);
    eprintln!();
    eprint!("Approve? [y/N] ");
    io::stderr().flush().ok();
    let mut line = String::new();
    if io::stdin().read_line(&mut line).is_err() {
        return (false, None, Some("input_error".to_string()));
    }
    let answer = line.trim().to_ascii_lowercase();
    if answer == "y" || answer == "yes" {
        (true, Some("cli_user".to_string()), None)
    } else {
        (false, None, Some("denied_by_user".to_string()))
    }
}
